### ISSUE TITLE
Color “simplification” screwup: Color groups are now incompatible with most other games and mods

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest_game#2203]
WTF? Why did you remove the dye color groups? Now it's just `color_*`. OMG, I can't even …

A lot of mods and games used this `unicolor` and `excolor` scheme. More than you think. It's a well-defined set of many colors, also written down in the README file. It worked very well as LOTS of mods and games have used theses groups. This is great because having a STANDARD set of color group means it works the same everywhere. `unicolor` and `excolor` are very powerful and very important and they have worked for YEARS. ANY decent mod dealing with colors uses these groups. I have yet to find a game or mod that deliberately ignored this standard. Except Minetest Game right now …
The color groups are not random, they are even cited in the Dev Wiki as a group set with well-defined meanings:
https://dev.minetest.net/Groups/Shared_groups

It's safe to say `excolor` and `unicolor` is a de-facto standard. Now thanks to this absolutely brain-dead “simplification”, Minetest Game is now the first game diverting from this standard. Breaking all the things in the name of “simplification” is a dangerous and stupid route, and should be avoided.

For instance, dye color groups in MineClone 2 are (I mean: WERE) compatible, although some of the actual dyes differ. This is very important for crafting. Now this compability is broken for no apparent reason. Now to “fix” this, all mods and games must be changed in order to please MTG. Your assumption that it only affects 2 mods is wrong. It is simply wishful thinking to belive that all mods and games will now follow Minetest Game's path. That won't happen.

The `unicolor`, `excolor` groups were basically a standard of defining colors which everyone followed. Now if Minetest Game starts deliberately diverting from this standard, it will dilute the standard and make it useless.

Furthermore, you failed to reach out to modders and game makers for such a serious change. You completely underestimated the impact of this. You know, VanessaE ain't the only modder. ;-)

Also, you completely failed to properly define the meaning of the new color groups, so they cannot be used as any standard whatsoever. And you completely failed to address it in `game_api.txt`. Apparently you didn't even read `game_api.txt` as it still mentions `excolor` and `unicolor`.

Please revert https://github.com/minetest/minetest_game/pull/2186 and https://github.com/minetest/minetest_game/pull/2142 and what else has been done to cripple, err, I mean “simplify” the color system.

**Comment by SmallJoker** [comment]
This kind of report would be helpful to have before the PRs are merged.

**Comment by Ezhh** [comment]
> The color groups are not random, they are even cited in the Dev Wiki

Info here isn't official or maintained specifically by the devs. (game_api.txt, on the other hand, should have been updated.)

> Furthermore, you failed to reach out to modders and game makers 

PR open for well over a month. How much longer should it have waited? Github, whether you like it or not, is where PRs are made and discussed, so any failure is also on the part of modders and game makers who don't check here. It's your choice not to check of course, but it means you have to accept that sometimes a change might break something for you, and you'll have to open an issue after to resolve it. It's not as if dyes can't be adjusted before the next release, so there's really no need for insults or dramatics. 

It would help if your provide examples of mods this breaks and so on so this can be assessed properly.

**Comment by tenplus1** [comment]
Out of every mod I checked only two used unicolor_ groups in which one was patched and the other isnt' necessary to work.

**Comment by Wuzzy2** [comment]
> The simplification was a bad idea, it did not add anything to the game and should simply be related.

Oops, I mean “reverted”.

**Comment by tenplus1** [comment]
Extensive list, you can remove bakedclay though cause I know that I dont use unicolor at all there...

I still say that anything that uses the base 15 colours should use the dye,colour_* grouping and anything more should look into supporting unifieddyes.

**Comment by Ezhh** [comment]
Since bakedclay belongs to and is maintained by @tenplus1, and doesn't use these groups, which bakedclay are you talking about?

Are the MTG-based games using the groups for anything, apart from having them specified in the copied version of the dye mod?

Other games on your list (for example, BFD) are not maintained, so quite likely to break for many other reasons. 

> Just hoping that eventually a modder will complain while not doing any research on your own?

Silly wuzzy. It's obvious we did look into this. I also spent some time slamming the color groups through searches in my browser. Really, not many results outside of the dye mod itself.

**Comment by paramat** [comment]
I'll update game_api.txt.
The PRs had been open a long time, the issue that discussed removing these groups open for much longer. Also we decided a little mod breakage is worth it to reduce the issues and bugs caused, so i made a news forum post to warn. MTG 5.0.0 is a good time to make such a breakage.
So i disagree with a revert even if there is a little disruption.

**Comment by Wuzzy2** [comment]
What's the point of this change if **nothing** of value is added? It's just a different set of names and has absolutely **zero** benefit. There has to be a stronger justification to break things. “It looks better” is not one. There's nothing wrong with the old groups.

Well, even with the `dye` mod alone, it's still breakage. The simple reason is: group-based crafting recipes. Those are (mostly) independent of a particular game or mod. All `unicolor` recipes are now worthless as well. My point is: These groups have worked just fine for years. This is the most pointless breakage ever.

Also, did I say that this breakage is 100% pointless and totally avoidable? The few minor bugs which this might have fixed are fixable in a non-disruptive manner.

The reason why I'm so angry is because of a singe game, you are forcing the entire Minetest community to adapt to this change for absolutely no good reason.
I will tell you what will happen if this change stays: A few modders will update, a few other modders won't (for various reasons).
At the end, we will end up with two incompatible sets of color groups, thus sabotaging any attempt to standardize colors even more.

Which leads to “group proliferation”. I.e. there are lots of different group names which are all slightl different, but actually mean the same thing. I know that groups are ultimately a mod/game thing, but we should at least *try* to keep things consistent.

>  Even the very experienced modder VanessaE recommended removing those groups.

I can not find any evidence to support this claim. And VanessaE isn't the only experienced modder around here …

**Comment by paramat** [comment]
See https://github.com/minetest/minetest_game/issues/1940#issuecomment-359246388
Also please read that whole thread, that somewhat explains why we did this.
Also related https://github.com/minetest-mods/technic/issues/370
The news forum post has no replies yet unfortunately.

Input from more modders on this would be appreciated. The PR was open for a month with no objections, now it's been merged it will likely come to the attention of more modders so hopefully we'll get more feedback. We're in dev and plenty of time to re-assess if necessary.

**Comment by paramat** [comment]
Hmm the forum post has no replies and 227 views, i actually expected more objection than this :)

**Comment by Wuzzy2** [comment]
`unicolor` is good because it is a relatively simple scheme with more than enough colors to use. Plus, it is very well established across mod AND game boundaries. I don't think you understand how powerful this actually is.

I really don't see the point of the new color scheme. It's not even well-defined, does not even offer more color options. It's basically just a rename that causes breakage for no reason. The colors the new color groups define are just the same. So basically you didn't even change anything, you just want to break things.

"Matching wool colors" is a pretty lame reason, since `unicolor` can model all wool colors just fine anyway.

> Mineclone2 is a self-contained game so is irrelevant.

No, it is NOT irrelevant. There is very well an use case for mods that work in multiple subgames, so having a standard that is established community-wide is very valuable and powerful.

If you think there is anything wrong with current groups, you should have discussed it BEFORE doing such extreme changes.

> Hmm the forum post still has no replies and 227 views, i actually expected more objection than this :)

That's not a valid argument.

**Comment by rubenwardy** [comment]
> No, it is NOT irrelevant. There is very well an use case for mods that work in multiple subgames, so having a standard that is established community-wide is very valuable and powerful.

I agree with this. Inter-game operability is important and some similarities makes porting much easier. Groups are a massive tool in this

**Comment by Ezhh** [comment]
> ... so having a standard that is established community-wide is very valuable and powerful.

No one disputed this. It's why I asked for examples of mods that are broken by this so we can do a real assessment of whether there are enough of them to make this worth reverting or adjusting. But when your list includes a mod that doesn't use the groups, it makes the list questionable at best.

I support closing this unless we get some meaningful real examples to review.

**Comment by Wuzzy2** [comment]
If you are looking for examples just scroll up a few posts … -_-

**Comment by Ezhh** [comment]
To repeat myself:

> But when your list includes a mod that doesn't use the groups, it makes the list questionable at best.

**Comment by Wuzzy2** [comment]
You are speaking in mysterious words.

**Comment by Ezhh** [comment]
Oh, I guess you were too busy raging to actually read the replies you were given by a dev who actually wanted to take you seriously.

Let's quote myself again.

> Since bakedclay belongs to and is maintained by @tenplus1, and doesn't use these groups, which bakedclay are you talking about?

That's been sitting unanswered for 11 days, so you certainly had time to explain its inclusion in your list. Until its inclusion in your list is explained, your list remains questionable. 

Also maybe help us to help you. The old system with so many groups was a mess. Despite your thoughts otherwise, cleaning up mess can be a very good thing. If that cleanup has actually broken something, drop the insults and give a simple explanation, then maybe we can find something that suits everyone.

**Comment by Wuzzy2** [comment]
OK, `bakedclay` is indeed no longer valid, I had an old version sitting around on my computer. Forget `bakedclay` then.

Anyway, as if that invalidates the rest of the list …

Note it is not possible for me to edit comments.

**Comment by Wuzzy2** [comment]
Also, how do you think that all mods and subgames ever made will update? Certainly a huge portion of MTG-derived games are now automatically incompatible, it's unlikely every single of them will just mindlessly follow MTG.
This is not about some possible breakage, this is about breakage you already caused. The breakage is that this just litters the MT modding sphere with more incompatible color grouping schemes. This sucks.
Thanks to MTG's deviation from the standard, the “mess” (as you call it) just becomes worse.
This is not helping anything.

You did not even try to preserve existing groups *at all*. Not even the slightest attempt of compability. You could, you know, at least try to use the `unicolor` group. But nooo, you just want to break things.
Oh, and you also threw the `bright` and `dark` options in `unicolor` under the bus. Those were useful!. So basically the new system is not simplified, it has been dumbed-down.

If you would only have wanted to nuke `basecolor` and `excolor`, I would have almost understood. But you want to destroy everything, unilaterally.

IMO the new system has just been hastily put together without putting much thought into it. Nor has the new scheme seen any serious discussion.

I don't like that attitude.

Do you really believe the entire MT community just follows whatever MTG dictates? It doesn't work that way. MTG is not the leader.



AAANYWAY. If the goal really is to nuke everything, it *definitely* should have seen more community involvement. It's better if everyone can agree on a simple color group standard than if every game and every mod just adds their own incompatible groups just because they feel like it. Imagine every modder would have the same attitude as you and ignore all other groups that exist. That would be utter chaos.

I recommend to revert the problematic commits, then start over, but this time, *do it right*.

**Comment by Wuzzy2** [comment]
Looks like you ninja'd a comment right after my comment. :D

> Again, a standard can be changed and that's what we're doing.

See, that's exactly what is wrong. MTG is NOT the leader. You just unilaterally decided to “change the standard”. But it doesn't work that way. Other games/mods disagree. If we take this attitude to its logical conclution, the result is an unholy mess of incompatible groups. Which means as a modder, you have to add many bullshit color groups just to make it work in multiple games.
Instead of having to set just use one color group scheme that worlds everywhere.

> Well it is

No. An argument from popularity is not an argument at all.

> "Are the MTG-based games using the groups for anything, apart from having them specified in the copied version of the dye mod?"

That's completely besides the point. The point is that every dye mod in these games should have *the same groups*. The mods might be “happy to update”, but the games probably not.
In other words, to ensure maximum compability as a modder, you have to combine the worst of two worlds:
1) The groups from the non-MTG games that are all the same
2) The groups from MTG that are different

Again: How do you think that mods should cope with this unilateral change to ensure maximum compability with most games (including MTG)?

**Comment by Ezhh** [comment]
After all your searching we're now down to four mods. The fact that four mods is all that's been found in this time that might now have compatibility issues between games (we're assuming they would not have other compatibility issues anyway), is a very small number of mods. There's been, as paramat pointed out, 6 years worth of these groups existing for them to make it into mods, but we have four, not even one a year. This isn't much breakage at all.

As for the games, them having different groups actually only matters if there are mods using them. That's why it's worth looking at how often the groups are actually used in mods.

> You could, you know, at least try to use the unicolor group. 

This is actually the possibility I was going to explore, but the dramatics and accusations don't motivate.

> I don't like that attitude.

You're funny, wuzzy.

**Comment by paanrama** [comment]
>MTG is NOT the leader.
I think that MTG is the leader, if not it would not have a lot of forks and variants, and not be used as the base for a lot of other games. If MTG is not the lider it can make all changes that it want without asking, and this discussion be irrelevant.

**Comment by paanrama** [comment]
>MTG is NOT the leader.

I think that MTG is the leader, if not it would not have a lot of forks and variants, and not be used as the base for a lot of other games. If MTG is not the lider it can make all changes that it want without asking, and this discussion be irrelevant.

**Comment by Wuzzy2** [comment]
Mods are not the whole story here.
Why do you keep dismissing the usage in games entirely? It's not good if games have two wildly-different, conflicting color schemes. That kind of defeats the point of groups.

What do you think is easier to achive?

1. EVERY SINGLE game in the list switches to whatever MTG thinks is best
2. MTG makes a small incremental change that is an improvement while at least trying to have some backwards-compability

Also, it has not been espablished at all why `unicolor` is bad. I think it's quite good. It offers a wide range of colors and is fairly generic and re-usable. The colors have some logic and system behind them. It *works*.

But the new color scheme is LESS flexible than unicolor. There is no genericism at all. You basically only added the wool colors. I mean, come on! This is not generic at all. There's dark green but no dark blue. There's pink and light gray, but no other bright colors. This doesn't even make sense.
This color system is so obviously biased towards MTG that it's useless for anything but mods for MTG. It can't even be re-used in other games if they change the colors even slightly.

If you really wanted to improve colors, you would have at least tried to have some system here. Like `unicolor` already offers.

I think the easiest and best simplification would be to drop `basecolor_` and `excolor_` as legacy while only leaving `unicolor_`. This is still a massive simplification AND also backwards-compatible.

With such a change, your “it's only 4 mods” argument would also actually be meaningful.

**Comment by paramat** [comment]
Unicolor describes brown as dark orange, and pink as light red, many people don't make this connection, i didn't for most of my life until i studied colur more.

The forum thread has been open almost a month now, still not any other complaints there or here. I must admit i expected more complaints.
